cpp/src/arrow/flight/sql/odbc/flight_sql/accessors/date_array_accessor_test.cc

```diff
@@ -60,19 +64,22 @@ TEST(DateArrayAccessor, Test_Date32Array_CDataType_DATE) {
 
   for (size_t i = 0; i < values.size(); ++i) {
     ASSERT_EQ(sizeof(DATE_STRUCT), strlen_buffer[i]);
-    tm date{};
 
-    int64_t converted_time = values[i] * 86400;
-    GetTimeForSecondsSinceEpoch(date, converted_time);
-    ASSERT_EQ((date.tm_year + 1900), buffer[i].year);
-    ASSERT_EQ(date.tm_mon + 1, buffer[i].month);
-    ASSERT_EQ(date.tm_mday, buffer[i].day);
+    ASSERT_EQ(expected[i].year, buffer[i].year);
+    ASSERT_EQ(expected[i].month, buffer[i].month);
+    ASSERT_EQ(expected[i].day, buffer[i].day);
   }
 }
 
 TEST(DateArrayAccessor, Test_Date64Array_CDataType_DATE) {
-  std::vector<int64_t> values = {86400000,  172800000, 259200000, 1649793238110,
-                                 345600000, 432000000, 518400000};
+  std::vector<int64_t> values = {86400000,  172800000,      259200000, 1649793238110,
+                                 0,         345600000,      432000000, 518400000,
+                                 -86400000, -17987443200000};
+  std::vector<tagDATE_STRUCT> expected = {
+      /* year(16), month(u16), day(u16) */
+      {1970, 1, 2}, {1970, 1, 3}, {1970, 1, 4}, {2022, 4, 12},  {1970, 1, 1},
+      {1970, 1, 5}, {1970, 1, 6}, {1970, 1, 7}, {1969, 12, 31}, {1400, 1, 1},
+  };
```

Comment:
Suggestion: could we add a test for a date with year before 1400? E.g., {1399, 2, 14}